src/storage/data_table.cpp

```diff
@@ -68,11 +69,18 @@ DataTable::SlotIterator &DataTable::SlotIterator::operator++() {
   common::SpinLatch::ScopedSpinLatch guard(&table_->blocks_latch_);
   // Jump to the next block if already the last slot in the block.
   if (current_slot_.GetOffset() == table_->accessor_.GetBlockLayout().NumSlots() - 1) {
-    ++block_;
-    // Cannot dereference if the next block is end(), so just use nullptr to denote
-    current_slot_ = {block_ == table_->blocks_.end() ? nullptr : *block_, 0};
+    if (num_advances_ > 0 || num_advances_ == SlotIterator::TILL_END) {
+      // Advance to the next block.
+      ++block_index_;
+      num_advances_ = num_advances_ == SlotIterator::TILL_END ? num_advances_ : num_advances_ - 1;
+      // Cannot dereference if the next block is end(), so just use nullptr to denote
+      current_slot_ = {block_index_ >= table_->blocks_.size() ? nullptr : table_->blocks_[block_index_], 0};
+    } else {
+      // Done advancing, time to give up.
+      current_slot_ = {nullptr, 0};
```

Comment:
The `nullptr` here is extremely opaque and makes it very hard to understand why this works given the updated constructor for `SlotIterator`.  We should use create a constant for `{nullptr, 0}` to have a clearer name for what this tuple-slot represents.